### PR TITLE
Faster fuzzy matching on user supplied parameters and paths

### LINKED ISSUE #373: Have fuzzy matching on user supplied parameters and paths
In the `prepare_run.py` when a user supplied parameters is not correct it would be very nice if the parameter is matched against all possible parameters and the closest one is given back to the user as a suggestion.

It could look like the following example:
```
Error 01: Receptor file not found

    |  ~/project_01/files/receptor.pdb

help= the folder 'project_01' does not live in '~' did you mean 'project-01'?
```

Discussed with @joaomcteixeira.

### PR DESCRIPTION
**Pull request by douweschulte**
Closes #373.

You are about to submit a new Pull Request. Before continuing make sure you read the [contributing guidelines](CONTRIBUTING.md) and you comply with the following criteria:

- [x] You have stick to Python. Talk with us before adding other programming languages to HADDOCK3
- [ ] Your PR is about CNS
- [x] Your code is well documented: proper docstrings and explanatory comments for those tricky parts
- [x] You structured the code into small functions as much as possible. You can use classes if there's a (state) purpose
- [x] code follows our coding style
- [x] You wrote tests for the new code
- [x] `tox` tests pass. *Run `tox` command inside the repository folder*
- [x] `-test.cfg` examples execute without errors. *Inside `examples/` run `python run_tests.py -b`*
- [x] PR does not add any *install dependencies* unless permission granted by the HADDOCK team
- [x] PR does not break licensing
- [ ] Your PR is about writing documentation for already existing code :fire:
- [ ] Your PR is about writing tests for already existing code :godmode:

---

Same as previous PR but now faster hopefully.

### CODE REVIEW

File: src/haddock/gear/prepare_run.py [hunk 1]
```diff
@@ -212,16 +215,24 @@ def validate_modules_params(modules_params):
...
+        all_parameters = \
+            set.union(set(extract_keys_recursive(defaults)),
+                      set(config_mandatory_general_parameters),
+                      set(non_mandatory_general_parameters_defaults.keys()),
+                      expandable_params)
+
+        diff = set(extract_keys_recursive(args)) - all_parameters
 
         if diff:
+            matched = fuzzy_match(diff, all_parameters)
```
Comment (joaomcteixeira) [inline src/haddock/gear/prepare_run.py:227]: `all_parameters` must be a list.

**Comment by joaomcteixeira** [comment]
Looks great, I am waiting for the `examples` tests. Once confirmed you can merge. I let you know.

**Review by joaomcteixeira — APPROVED** [review]
Looks great.
All tests pass now.
Messages are clear, and the evaluation is fast now. You can merge it when you wish.

**Comment by rvhonorato** [comment]
Would be cool if it could check if the path is a softlink and then display both (this is usually the cause of issue when running things in the hpc centers) and also check if the path you are referring is mounted somewhere else. 

Now it just handles typos, right?

**Comment by joaomcteixeira** [comment]
It handles typos on parameters, modules, and filenames. It does not check if the path is a  link. But that might be a good addition. I think it is best to have it in a new PR, though, not to overload this one.

**Pull request merged**